lib/aws/client.ex

```diff
@@ -194,12 +194,97 @@ defmodule AWS.Client do
     %{client | http_client: http_client}
   end
 
+  @doc """
+  Makes a HTTP request using the specified client.
+
+  `opts` may contain the keyword `:enable_retries?` that enables request retries on known
+  errors such as 500s.
+  The keyword `:retry_opts` can further supply the arguments: `:max_retries`, `:base_sleep_time` &
+  `:cap_sleep_time` that control the retry mechanism. This uses exponential backoff with jitter to
+  sleep in between retry attempts. Default values are:
+  `retry_opts: [max_retries: 10, base_sleep_time: 5, cap_sleep_time: 5_000]`
+
+  See "FullJitter" at: https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/
+
+  ### Examples
+
+      iex> AWS.Client.request(client, :post, url, payload, headers, options)
+      {:ok, %{status_code: 200, body: body}}
+
+      iex> AWS.Client.request(client, :post, url, payload, headers, enable_retries?: true)
+      {:ok, %{status_code: 200, body: body}}
+  """
   def request(client, method, url, body, headers, opts \\ []) do
+    # Pop off all retry-related options from opts, so they aren't passed to the HTTP client.
+    {enable_retries?, opts} = Keyword.pop(opts, :enable_retries?, false)
+    {retry_num, opts} = Keyword.pop(opts, :retry_num, 0)
+    {retry_opts, opts} = Keyword.pop(opts, :retry_opts, [])
+    # Defaults for retry_opts
+    retry_opts =
+      Keyword.merge([max_retries: 10, base_sleep_time: 5, cap_sleep_time: 5_000], retry_opts)
```

Comment:
Instead of `Keyword.merge/2`, we could use [`Keyword.validate!/2`](https://hexdocs.pm/elixir/main/Keyword.html#validate/2) with the defaults, so any other option will raise. The only restriction is that this is only available in `Elixir ~> 1.13`.

```suggestion
      Keyword.validate!(retry_opts, max_retries: 10, base_sleep_time: 5, cap_sleep_time: 5_000, retry_num: 0)
```

Reply:
Ah, I wasn't aware of that!
Do you think it makes sense to break backwards compat for this? I tried to work around that elsewhere.

Reply:
For the 1.0 version, I believe we should require Elixir 1.13, yeah. WDYT @onno-vos-dev ?

Reply:
@philss @onno-vos-dev bumping this conversation

Reply:
Hi sorry for the delay. Been a bit hectic for me lately. I'd personally vote against breaking backwards compatibility for this reason. I don't think the reason is strong enough in my eyes so I'd keep it as is for now. Not knowing exactly how the Elixir landscape looks like, I'd imagine a decent chunk of services still use 1.13 and this backwards compatibility breakage doesn't warrant abandoning those in my mind.